src/Authentication/Authentication/Helpers/HttpMessageFormatter.cs

```diff
@@ -0,0 +1,359 @@
+    private const string ColonSP = ": ";
+    private const string CRLF = "\r\n";
+    private const string CommaSeparator = ", ";
+
+    private const int DefaultHeaderAllocation = 2 * 1024;
+
+    private const string DefaultMediaType = "application/http";
+
+    private const string MsgTypeParameter = "msgtype";
+    private const string DefaultRequestMsgType = "request";
+    private const string DefaultResponseMsgType = "response";
+
+    // Set of header fields that only support single values such as Set-Cookie.
+    private static readonly HashSet<string> _singleValueHeaderFields = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+    {
+      "Cookie",
+      "Set-Cookie",
+      "X-Powered-By",
+    };
+
+    // Set of header fields that should get serialized as space-separated values such as User-Agent.
+    private static readonly HashSet<string> _spaceSeparatedValueHeaderFields = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+    {
+      "User-Agent",
+    };
+
+    // Set of header fields that should not get serialized
+    private static readonly HashSet<string> _neverSerializedHeaderFields = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+    {
+      "SdkVersion",
+      "FeatureFlag",
+      "Authorization",
+      "Cache-Control",
+      "Transfer-Encoding",
+      "Duration",
+      "Strict-Transport-Security",
+      "Date"
+    };
+
+    private bool _contentConsumed;
+    private Lazy<Task<Stream>> _streamTask;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="HttpMessageContent"/> class encapsulating an
+    /// <see cref="HttpRequestMessage"/>.
+    /// </summary>
+    /// <param name="httpRequest">The <see cref="HttpResponseMessage"/> instance to encapsulate.</param>
+    public HttpMessageFormatter(HttpRequestMessage httpRequest)
+    {
+      HttpRequestMessage = httpRequest ?? throw new ArgumentNullException("httpRequest");
+      Headers.ContentType = new MediaTypeHeaderValue(DefaultMediaType);
+      Headers.ContentType.Parameters.Add(new NameValueHeaderValue(MsgTypeParameter, DefaultRequestMsgType));
+
+      InitializeStreamTask();
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="HttpMessageContent"/> class encapsulating an
+    /// <see cref="HttpResponseMessage"/>.
+    /// </summary>
+    /// <param name="httpResponse">The <see cref="HttpResponseMessage"/> instance to encapsulate.</param>
+    public HttpMessageFormatter(HttpResponseMessage httpResponse)
+    {
+      HttpResponseMessage = httpResponse ?? throw new ArgumentNullException("httpResponse");
+      Headers.ContentType = new MediaTypeHeaderValue(DefaultMediaType);
+      Headers.ContentType.Parameters.Add(new NameValueHeaderValue(MsgTypeParameter, DefaultResponseMsgType));
+
+      InitializeStreamTask();
+    }
+
+    private HttpContent Content
+    {
+      get { return HttpRequestMessage != null ? HttpRequestMessage.Content : HttpResponseMessage.Content; }
+    }
```

Comment:
Q: Is this a convention of some sort?

Reply:
The formatter can format both requests and responses, so here is to return whichever is available.

Reply:
I can't think of a scenario where the consumer of this code would ask for `.Content` property without knowing whether they are going to get the contents of request or response. That sounds like an arbitrary behavior. But maybe I am missing an important detail.

Do you mind clarifying with a call flow for me to better understand the coupling of "`Content` can return response content if an only if request is null"?

If this behavior is not intended, I would suggest having two different properties, e.g. `RequestContent` and `ResponseContent`.

Reply:
HttpMessageFormatter, can format both requests and responses. It has two constructors, one takes a request and the other takes a response. So when a Request is passed in (this field is private though), HttpContent will point to the the Request content.